code/object/objcollide.cpp

```diff
@@ -98,137 +98,184 @@ int reject_due_collision_groups(object *A, object *B)
 
-	if ( !(A->flags[Object::Object_Flags::Collides]) ) return;		// This object doesn't collide with anything
-	if ( !(B->flags[Object::Object_Flags::Collides]) ) return;		// This object doesn't collide with anything
+    // Make sure you're not checking a parent with it's kid or vicy-versy
+    if ( reject_obj_pair_on_parent(A,B) ) return false;
 
-	if ((A->flags[Object::Object_Flags::Immobile]) && (B->flags[Object::Object_Flags::Immobile])) return;	// Two immobile objects will never collide with each other
-	
-	// Make sure you're not checking a parent with it's kid or vicy-versy
-	if ( reject_obj_pair_on_parent(A,B) ) {
-		return;
-	}
+    //else
+    return true;
+}
 
-	Assert( A->type < 127 );
-	Assert( B->type < 127 );
-
-	ctype = COLLISION_OF(A->type,B->type);
-	switch( ctype )	{
-	case COLLISION_OF(OBJ_WEAPON,OBJ_SHIP):
-		swapped = 1;
-		check_collision = collide_ship_weapon;
-		break;
-	case COLLISION_OF(OBJ_SHIP, OBJ_WEAPON):
-		check_collision = collide_ship_weapon;
-		break;
-	case COLLISION_OF(OBJ_DEBRIS, OBJ_WEAPON):
-		check_collision = collide_debris_weapon;
-		break;
-	case COLLISION_OF(OBJ_WEAPON, OBJ_DEBRIS):
-		swapped = 1;
-		check_collision = collide_debris_weapon;
-		break;
-	case COLLISION_OF(OBJ_DEBRIS, OBJ_SHIP):
-		check_collision = collide_debris_ship;		
-		break;
-	case COLLISION_OF(OBJ_SHIP, OBJ_DEBRIS):
-		check_collision = collide_debris_ship;
-		swapped = 1;
-		break;
-	case COLLISION_OF(OBJ_ASTEROID, OBJ_WEAPON):
-		check_collision = collide_asteroid_weapon;
-		break;
-	case COLLISION_OF(OBJ_WEAPON, OBJ_ASTEROID):
-		swapped = 1;
-		check_collision = collide_asteroid_weapon;
-		break;
-	case COLLISION_OF(OBJ_ASTEROID, OBJ_SHIP):
-		check_collision = collide_asteroid_ship;
-		break;
-	case COLLISION_OF(OBJ_SHIP, OBJ_ASTEROID):
-		check_collision = collide_asteroid_ship;
-		swapped = 1;
-		break;
-	case COLLISION_OF(OBJ_SHIP,OBJ_SHIP):
-		check_collision = collide_ship_ship;
-		break;	
-	
-	case COLLISION_OF(OBJ_BEAM, OBJ_SHIP):
-		if(beam_collide_early_out(A, B)){
-			return;
-		}
-		check_collision = beam_collide_ship;
-		break;
+using CollisionChecker = int (*)( obj_pair * );
 
-	case COLLISION_OF(OBJ_BEAM, OBJ_ASTEROID):
-		if(beam_collide_early_out(A, B)){
-			return;
-		}
-		check_collision = beam_collide_asteroid;
-		break;
+/**
+ * @brief set type of collision based on types of objects colliding
+ * @param A
+ * @param B
+ * @return true if no further collision resolution is expected and colling function can return early
+ * @post check_collision != nullptr || return false
+ *
+ * This function was extracted from duplicate code in obj_add_pair and obj_collide_pair
+ */
+static bool check_collision_type(object *A, object *B, CollisionChecker& check_collision, int& swapped)
+{
+    Assert(A != nullptr);
+    Assert(B != nullptr);
+    Assert( A->type < 127 );
+    Assert( B->type < 127 );
 
-	case COLLISION_OF(OBJ_BEAM, OBJ_DEBRIS):
```

Comment:
The original switch case in obj_add_pair was not symmetric for beam weapons.
(It covered COLLISION_OF(OBJ_BEAM, x) but not COLLISION_OF(x, OBJ_BEAM) )

Does anyone know if that is intended or even necessary?
It seems buggy to me.

Reply:
It was probably intentional. 
For every pair (x, y) an equivalent pair (y, x) will exist; technically you only need to check one combination (you'll note that for those duplicate cases, a "swapped" flag is set, which will swap the A and B object references around before actually resolving the collision). 
One possible performance optimization could be to try and cull those duplicate pairs; however, I do not know if the overhead of performing that cull is larger than rejecting those pairs during the resolution phase, which is what the code is currently doing.

Reply:
Maybe it is by construction, that beam objects will always be at the end of the array or something like that, and thus collisions will only ever be in that order.

But why is do the symetric cases exist for other combinations like WEAPON + x?
Does anyone know why Swifty added them in his (mostly) duplicate code?
His commit message doesn't really tell anything.

Reply:
It seems that obj_add_pair is actually dead code.
It is only called in obj_add_pairs in object.cpp, which is never called.

So instead of unifying the function, I'll just delete one.